# 1. Data Cleaning and Feature Engineering

This section covers the initial steps of data cleaning and feature engineering.  
The primary goal is to understand the raw data, handle preliminary issues, and create new, more useful features from the existing ones.

## 1.1 Initial Data Inspection and Handling

The first step consisted of a preliminary inspection of the raw dataset.  
To better understand the content and structure of each column, the DataFrame was exported to an Excel file for quick visual inspection.

However, this led to an immediate technical issue.

### Problem: Excel Formula Injection

Some text fields begin with special characters such as `=`, `+`, `-`, or `@`.

When opened in Excel, these are interpreted as formulas, which may corrupt the data or generate errors.

To prevent this issue, a helper function was implemented to prepend a single quote (`'`) to any string starting with these characters, forcing Excel to interpret them as plain text.

In [2]:
import pandas as pd
import numpy as np

file_path = 'data/apple_jobs.csv' 

In [3]:
# 3. Leggere il file CSV e caricarlo in un DataFrame Pandas
try:
    df = pd.read_csv(file_path)
    print("File caricato con successo!")
except FileNotFoundError:
    print(f"Errore: il file '{file_path}' non è stato trovato.")

File caricato con successo!


In [4]:
def escape_excel_formula(x):
    if isinstance(x, str) and x.startswith(('=', '+', '-', '@')):
        return "'" + x
    return x

df_excel = df.map(escape_excel_formula)

# Esportiamo in Excel
output_path = 'data/apple_jobs_EXCEL_visivo.xlsx'
df_excel.to_excel(output_path, index=False)  # index=False evita di esportare l'indice come colonna

print(f"File Excel creato con successo: {output_path}")

File Excel creato con successo: data/apple_jobs_EXCEL_visivo.xlsx


## 1.2 Standardizing Job Titles into Role Categories

The `title` column is a free-text field containing multiple variations of job roles.

The objective is to create a new standardized column called `Role_Category` to group the most common roles into:

- **SD** → Software Developer  
- **SE** → Software Engineer  
- **OTHER** → All remaining roles

### Attempt 1: Strict Matching

The first approach used regular expressions to match full words such as:

- developer
- engineer

Word boundaries (`\b`) were used to ensure exact matching.

In [15]:
# Create lowercase version for matching
df['Title_clean'] = df['title'].str.lower()

cond_sd_v1 = df['Title_clean'].str.contains(r'\bdev(?:eloper)?\b', na=False)
cond_se_v1 = df['Title_clean'].str.contains(r'\beng(?:ineer)?\b', na=False)

df['Role_Category'] = np.select(
    [cond_sd_v1, cond_se_v1],
    ['SD', 'SE'],
    default='OTHER'
)

df['Role_Category'].value_counts()

Role_Category
SE       1380
OTHER     672
SD         96
Name: count, dtype: int64

### Observation

This approach left a high number of roles classified as **OTHER**.

After inspecting these titles, it became clear that many contained abbreviations such as "eng" or partial words like "dev".

### Attempt 2: Flexible Matching

To improve the classification, partial string matching was used instead:

- "dev"
- "eng"

In [17]:
cond_sd_v2 = df['Title_clean'].str.contains(r'dev', na=False)
cond_se_v2 = df['Title_clean'].str.contains(r'eng', na=False)

df['Role_Category'] = np.select(
    [cond_sd_v2, cond_se_v2],
    ['SD', 'SE'],
    default='OTHER'
)

df['Role_Category'].value_counts()

Role_Category
SE       1454
OTHER     488
SD        206
Name: count, dtype: int64

## Reflection

The second approach did not substantially reduce the number of unclassified roles.

However, this raises important analytical questions:

1. Is the SD vs SE distinction truly meaningful?
2. Does the dataset reveal structural differences between these roles?
3. Will this feature uncover meaningful patterns, or is it an oversimplification?

I'm starting to think that my analysis is too generic. Then, what does it really mean to be a developer rather than an engineer? What weight could this distinction carry in our analysis?

## Pure Study of the Variables

Let's pause and take a breath… we are mixing the two methods. In the first analysis, we should **not modify any variable**, but rather understand the meaning of each one before considering the client’s requirements.  

So, if I were to start over, I would use Python code to **visualize the features appropriately and in a readable way**, generating the file **apple_jobs_EXCEL_visivo**. From here, there are two ways to continue:

1. Extract every value of the variable (in this case textual) and have the AI explain in detail what it means. This is because it is difficult to manually analyze almost 2,000 unique texts.  
2. Get a sense of what the variable is trying to represent by sampling only some of the texts.  

I will proceed using the second method first, and then the first one, and compare which approach is more useful by evaluating both speed and efficiency.  

### Manual Description of Variables

Below is a description of the variables:

1. **title:** It is noticeable that the titles start with standard definitions like “Software Engineer,” and then move to more detailed and complex definitions. They are all engineers, but they perform different tasks and have different levels of experience. One could consider subdividing these engineers into more homogeneous macro-categories. For example: front-end software engineers, back-end SE, senior iOS SE, Java SE, etc. To do this without AI, we would need to find keywords for each segmentation (for example: java, iOS, back, front, etc.) and automate the classification by assigning a standard name.  

2. **location:** First, we check how many unique values exist. If there are different cities, the problem is standardizing the nomenclature. We can automate this with a Python function that, each time it encounters a new field, compares it with the others stored in memory (as a vector of words). If it mostly matches one of them (for example, only one or two words differ), it assigns the standard name of that location; otherwise, it stores it in memory as a new value.  

3. **minimum_qual, preferred_qual, responsibilities, education & experience:** These variables are really difficult to standardize into data without using AI. They describe the minimum and preferred requirements for the job, the responsibilities of the employee, and the education and experience required.  

## Objectives of the Work

So, we understand that the dataset concerns job openings in certain specific cities. Now the questions we ask ourselves are:

1. What information can we extract from the data?  
2. Who could benefit from our investigation?  
3. What does the client commissioning the study want to know?  

From the data, we could understand how jobs are distributed geographically, and how the location affects the difficulty and access requirements for a job as a software engineer. For example, this investigation could serve someone looking for a job who doesn’t know where to relocate. Or it could help other companies looking for employees in that area to understand the competition and align themselves. Variables such as proximity to important universities or work centers could also be included. At the moment, no other practical uses for this dataset come to my mind.

---

## REGEX ANALYSIS

Our **regex-based classification** attempt, designed to assign each job title either the SE or SD label, proved insufficient, likely because the values in the **Role_Category** column are highly heterogeneous. A simple pattern-matching approach cannot capture such variability, as evidenced by the fact that over 500 records were classified as OTHER. https://regex101.com/ Important tool.

Follows some examples, taken from the first rows of the dataset, that can easily prove our point:

In [4]:
with pd.option_context('display.max_colwidth', None):
    i=1
    rows_num = 6
    for val in df['minimum_qual'].head(rows_num):
        print()
        print(str(i) + '.   ' + val)
        i += 1


1.   2-5 years experience supporting or deploying application software for internal and external customers.
Experience configuring or maintaining commercial enterprise application software packages (for example: inventory systems, manufacturing systems, PDM, ERP, LIMS, etc…)
Experience providing first level software support to end users and issue triage
Proven problem solving and debugging skills
Comfortable with modern IT tech-stack concepts
Ability to navigate in a Linux/Unix environment
Able to train small and large groups
Able to write and execute basic database queries
You will be self directed, analytical, and work well in a team environment
Be a strong advocate for improving processes and a clear communicator of new ideas
Strong verbal interpersonal skills
Highly organized, ability to juggle multiple priorities at a time
Ability to quickly learn new software applications

2.   KEY QUALIFICATIONS
A genuine passion for fixing software and hardware issues.
Strong knowledge of proc

## DIGRESSION - NLP Model (SpaCy)
Before continuing with the regex-based approach, I want to try this method: a model in the library SpaCy that can analyze each text and extract patterns. The full documentation can be found here https://spacy.io/usage#installation


### What is spaCy and what are we doing with it?

**spaCy** is a Python library for **Natural Language Processing (NLP)**, meaning it helps us analyze and understand text written in natural language.

### Loading the model
When we write:
```python
nlp = spacy.load("en_core_web_sm")
```
We are loading a **pre-trained model** — an "intelligence" already trained on millions of English texts that understands the structure of language.

### Analyzing the text
When we write:
```python
doc = nlp(text)
```
spaCy **analyzes the text** and automatically performs several operations:
- splits the text into words (**tokenization**)
- recognizes the grammatical role of each word (verb, noun, adjective...)
- identifies **Named Entities (NER)** — understanding whether a word refers to an organization, a date, a location, a person, etc.

### Reading the results
When we write:
```python
for ent in doc.ents:
    print(ent.text, '-', ent.label_)
```
We are printing the **recognized entities**. This is why it returned things like `2-5 years - DATE` or `Linux - GPE`.

In [5]:
import spacy

nlp = spacy.load("en_core_web_sm")
num = 5
i=0
while i < num:
    #prende il campo testuale
    testo = df['minimum_qual'][i]
    #si analizza
    doc = nlp(testo)
    #stampa quello che ha trovato
    print()
    print(str(i) + '. ')
    i=i+1
    for ent in doc.ents:
        print(ent.text, '-', ent.label_)


0. 
2-5 years - DATE
PDM - ORG
ERP - ORG
LIMS - ORG
first - ORDINAL
Linux - GPE

1. 
CPU/Bus - ORG
Intel - ORG
EFI - ORG
DDR - ORG
Battery - LOC
PCB - ORG
SMD - ORG
Kernal - PERSON

2. 
Java - PERSON
at least one - CARDINAL
Bash (Ruby/Python/Go - ORG
Cassandra - PERSON
ZooKeeper - ORG
Kafka - PERSON

3. 
Java - PERSON
at least one - CARDINAL
Bash (Ruby/Python/Go - ORG
Cassandra - PERSON
ZooKeeper - ORG
Kafka - PERSON

4. 
Sensor Systhigh - PERSON
Controls, Statistical Inference - ORG
Performance Computing - PERSON
Strong - GPE
linear - ORG


As we can see in the previous results, the model is unable to extract useful information. It classified Java as a person and Linux as a location. In conclusion, we can say that this model is not appropriate for the kind of data we have.
There is another approach we can use that is not based on regex. It finds specific words contained in a list that we create.

---

## KEYWORD MATCHING

In [6]:
import re

skill_list = [
    # Programming languages
    'python', 'java', 'sql', 'r', 'scala', 'bash', 'ruby', 'go', 'c++', 'c#', 'swift', 'kotlin',
    # Machine learning / AI
    'machine learning', 'deep learning', 'nlp', 'computer vision', 'neural network',
    'tensorflow', 'pytorch', 'scikit-learn', 'keras',
    # Data
    'pandas', 'numpy', 'spark', 'hadoop', 'kafka', 'tableau', 'power bi',
    # Cloud
    'aws', 'azure', 'gcp', 'google cloud',
    # DevOps / tools
    'docker', 'kubernetes', 'git', 'linux', 'unix', 'agile', 'scrum',
    # Database
    'mysql', 'postgresql', 'mongodb', 'oracle', 'redis', 'cassandra',
    # Enterprise
    'erp', 'sap', 'salesforce'
]

def extract_skills(text):
    if isinstance(text, str):
        text_lower = text.lower()
        found = []
        for skill in skill_list:
            if ' ' in skill:
                if skill in text_lower:
                    found.append(skill)
            else:
                if re.search(r'\b' + re.escape(skill) + r'\b', text_lower):
                    found.append(skill)
        return found
    return []

df['skills_found'] = df['minimum_qual'].apply(extract_skills)

# verifica
print(df['skills_found'].head(40))

0                                    [linux, unix, erp]
1                                                [unix]
2      [python, java, bash, ruby, go, kafka, cassandra]
3      [python, java, bash, ruby, go, kafka, cassandra]
4                                    [machine learning]
5                                                    []
6                                               [swift]
7            [java, scala, swift, spark, kafka, oracle]
8                                               [swift]
9                                                    []
10                                              [swift]
11                              [java, sql, kafka, git]
12                                                   []
13                                                   []
14                            [python, computer vision]
15                         [python, ruby, linux, agile]
16                                              [swift]
17              [java, hadoop, kafka, redis, cas


This approach looks way better. Almost every row contains different values, but it is still not fully usable. There are many rows that do not contain any value, not because these jobs have no minimum qualifications, but because our list lacks a proper coverage of all possible skills. We can therefore choose between two paths: **expanding our dictionary** by adding more entries, or **switching to an LLM-based method**.

## Expanding the dictionary

Let's visualize and analyze only rows that have less then 3 values in it.

In [7]:
i = 0
for idx, row in df.iterrows():
    if len(row['skills_found']) < 3:
        i += 1
print('Totale righe con meno di 3 skill:', i)

Totale righe con meno di 3 skill: 1563


As we can see we have **1563** rows out of almost **2200** with less then 3 skills. That's obviously a **big problem**. Trying to resolve this problem, because we cannot analyze all 1563 rows, let's try to **introduce new skills** in the dictionary analyzing just the first 54 rows.

Let's print these rows:

In [23]:
i = 0
for idx, row in df.iterrows():
    if len(row['skills_found']) < 3:
        i += 1
        print(row['minimum_qual'])
        print()
        print(row['skills_found'])
        print()
        print()
        if i==5:
            break
print('Totale righe con meno di 3 skill:', i)

KEY QUALIFICATIONS
A genuine passion for fixing software and hardware issues.
Strong knowledge of processor architecture (CPU/Bus/subsystems/peripherals/clocks/power states).
Strong understanding of Intel processors, EFI, PCI, OS X/Unix.
Extensive knowledge of hardware platform (Power, DDR, storage, Battery, PCB, SMD).
Previous Experience with Kernal and OS Drivers
Understanding of on-chip debug and trace facilities. Exposure to hardware logic analyzers, oscilloscopes.
Solid C programming skills.
Understanding ARM/x86 assembly languages for debugging.
Previous experience with platform bring-up.
Demonstrated creative and critical thinking capabilities.
Excellent collaborative skills including written and verbal communication.

['unix']


Demonstrated experience integrating full-system autonomy stacks including developing interfaces to enable sub-system verification before field testing
Technical proficiency in multiple of: Perception, Machine Learning, Sensor Systhigh ems, Controls, Sta

I had this output analyzed by an LLM (Claude), which suggested adding the following new values:

- **Mobile / Apple ecosystem:** `ios`, `macos`, `objective-c`, `xcode`
- **Hardware / Low-level:** `arm`, `x86`, `assembly`, `firmware`, `embedded`
- **Architecture / Systems:** `distributed systems`, `microservices`, `rest`, `api`
- **Other:** `gui`

In [28]:
skill_list = [
    # Programming languages
    'python', 'java', 'sql', 'r', 'scala', 'bash', 'ruby', 'go', 'c++', 'c#', 'swift', 'kotlin',
    # Machine learning / AI
    'machine learning', 'deep learning', 'nlp', 'computer vision', 'neural network',
    'tensorflow', 'pytorch', 'scikit-learn', 'keras',
    # Data
    'pandas', 'numpy', 'spark', 'hadoop', 'kafka', 'tableau', 'power bi',
    # Cloud
    'aws', 'azure', 'gcp', 'google cloud',
    # DevOps / tools
    'docker', 'kubernetes', 'git', 'linux', 'unix', 'agile', 'scrum',
    # Database
    'mysql', 'postgresql', 'mongodb', 'oracle', 'redis', 'cassandra',
    # Enterprise
    'erp', 'sap', 'salesforce',
    # Mobile / Apple ecosystem
    'ios', 'macos', 'objective-c', 'xcode',
    # Hardware / Low-level
    'arm', 'x86', 'assembly', 'firmware', 'embedded',
    # Architecture / Systems
    'distributed systems', 'microservices', 'rest', 'api',
    # Other
    'gui'
]


def extract_skills(text):
    if isinstance(text, str):
        text_lower = text.lower()
        found = []
        for skill in skill_list:
            if ' ' in skill:
                if skill in text_lower:
                    found.append(skill)
            else:
                if re.search(r'\b' + re.escape(skill) + r'\b', text_lower):
                    found.append(skill)
        return found
    return []

df['skills_found_updated'] = df['minimum_qual'].apply(extract_skills)

# verifica
print(df['skills_found_updated'].head(40))

0                                    [linux, unix, erp]
1                            [unix, arm, x86, assembly]
2     [python, java, bash, ruby, go, kafka, cassandr...
3     [python, java, bash, ruby, go, kafka, cassandr...
4                                    [machine learning]
5                                                    []
6                 [swift, ios, macos, objective-c, gui]
7       [java, scala, swift, spark, kafka, oracle, ios]
8                 [swift, ios, macos, objective-c, gui]
9                                                    []
10                                 [swift, objective-c]
11    [java, sql, kafka, git, microservices, rest, api]
12                                [distributed systems]
13                                [distributed systems]
14          [python, computer vision, ios, objective-c]
15               [python, ruby, linux, agile, firmware]
16                                 [swift, objective-c]
17              [java, hadoop, kafka, redis, cas

The results look more solid than before. Let's now check how many rows still contain fewer than 3 extracted skills.

In [29]:
i = 0
for idx, row in df.iterrows():
    if len(row['skills_found_updated']) < 3:
        i += 1
print('Totale righe con meno di 3 skill:', i)

Totale righe con meno di 3 skill: 1306


We notice that the situation has not changed significantly. There are still more than 1300 rows with fewer than 3 extracted skills. We successfully reduced the empty records by about 200 units, but this is **not enough** to consider the **keyword matching method** valid for our specific use case.

## LLM ANALYSIS

As a last resort, we want to use an **LLM** such as **ChatGPT, Gemini or Claude** to read each entry and extract useful information. This method is not recommended for datasets with a large number of rows, but in our case, with approximately 2,200 records, it is perfectly feasible.

In [ ]:
from openai import OpenAI

client = OpenAI(api_key="YOUR_API_KEY_HERE")

def extract_skills_llm(text):
    if not isinstance(text, str):
        return []
    
    response = client.chat.completions.create(
        model="gpt-4o-mini",  # economico e veloce
        messages=[
            {
                "role": "system",
                "content": "You are a skill extractor. Given a job posting, extract only the technical skills, tools, and technologies mentioned. Return them as a comma-separated list, nothing else."
            },
            {
                "role": "user",
                "content": text
            }
        ]
    )
    
    result = response.choices[0].message.content
    return [s.strip().lower() for s in result.split(',')]

# testa su una riga
print(extract_skills_llm(df['minimum_qual'][0]))

['linux', 'unix', 'inventory systems', 'manufacturing systems', 'pdm', 'erp', 'lims', 'database queries']


https://developers.openai.com/api/docs/guides/structured-outputs follow this for prompt engineering